README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# whatthreewords

<!-- badges: start -->

[![R-CMD-check](https://github.com/DavidASmith/whatthreewords/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/DavidASmith/whatthreewords/actions/workflows/R-CMD-check.yaml)
[![CRAN
status](https://www.r-pkg.org/badges/version/whatthreewords)](https://CRAN.R-project.org/package=whatthreewords)
<!-- badges: end -->

The whatthreewords package supports working with the
[what3words](https://what3words.com/) API from R. what3words has
partitioned the surface of the earth into 3m x 3m squares, each of which
can be identified by three words. These are conventionally styled with
three slashes at the beginning. For example, the peak of the Great
Pyramid of Giza is located by the three words
[///ballots.height.silks](https://what3words.com/ballots.height.silks).

You can use the package to determine the three word address for any
coordinates, or return the coordinates for any valid three word
combination.

## Installation

You can install the development version of whatthreewords from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("DavidASmith/whatthreewords")
```

## Authentication

The what3words API requires a key for authentication. You can register
for a key at <https://developer.what3words.com/public-api>.

You must then set the `WTW_API_KEY` environment variable to hold your
key. For example…

``` r
Sys.setenv(WTW_API_KEY = "MYKEY")
```

## Examples

Here are a few examples of what you can do with the package. For more
details, see the vignette ‘Using whatthreewords’.

### Get a what3words address from coordinates

Use `words_from_coords` to get a what3words location for a given
latitude and longitude.

``` r
library(whatthreewords)

words_from_coords(lat = 51.5095, 
                  lon = -0.1266)
#> [1] "lamp.inner.dent"
```

This is vectorised over `lat` and `lon`.

``` r
words_from_coords(lat = c(53.3703, 53.41145, 53.3096), 
                  lon = c(-1.47119, -1.500204, -1.478715))
#> [1] "elite.icon.levels" "weep.stands.shack" "soon.belt.owls"
```

## Get coordinates from a what3words location

`coords_from_words` returns the coordinates for a what3words location.

``` r
coords_from_words("hours.flesh.petal")
#> Warning: ERROR: QuotaExceeded: Quota Exceeded. Please upgrade your usage plan,
#> or contact support@what3words.com
#>      lat lon
#> [1,]  NA  NA
```

## Alternative packages

There is an alternative R client for what3words available on
[CRAN](https://CRAN.R-project.org/package=threewords) ([Github
repo](https://github.com/Ironholds/threewords)). However, this has not
been updated (at time of writing) for several years and I couldn’t get
it to work. I needed something fairly quickly so wrote this from
scratch.


NEWS.md
# whatthreewords (development version)

# whatthreewords 0.1.3

* Apply CRAN review comments:
    * Use only undirected quotation marks in the description text.

# whatthreewords 0.1.2

* Apply CRAN review comments:
    * Software naming conventions.
    * Add link to description.

# whatthreewords 0.1.1

* Fixed `get_api_key()` example exposing API key in documentation.

# whatthreewords 0.1.0

* Initial release.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(whatthreewords)

test_check("whatthreewords")


tests/testthat/test-words_from_coords.R
test_that("API returns words from coordinates (single)", {
  skip_if(identical(Sys.getenv("WTW_API_KEY"), ""),
          message = "API key environment variable not set.")
  expect_identical(words_from_coords(lat = 51.5095, lon = -0.1266),
                   "lamp.inner.dent")
})

test_that("API returns words from coordinates (vectorised)", {
  skip_if(identical(Sys.getenv("WTW_API_KEY"), ""),
          message = "API key environment variable not set.")
  expect_identical(words_from_coords(lat = c(53.3703, 53.41145, 53.3096),
                                     lon = c(-1.47119, -1.500204, -1.478715)),
                   c("elite.icon.levels", "weep.stands.shack", "soon.belt.owls")
  )
})

test_that("Full details are returned (single)", {
  skip_if(identical(Sys.getenv("WTW_API_KEY"), ""),
          message = "API key environment variable not set.")
  full_details <- words_from_coords(lat = 51.5095,
                                    lon = -0.1266,
                                    full_details = TRUE)
  expect_type(full_details, "list")
  expect_identical(full_details$words, "lamp.inner.dent")
})


test_that("Full details are returned (vectorised)", {
  skip_if(identical(Sys.getenv("WTW_API_KEY"), ""),
          message = "API key environment variable not set.")
  full_details <- words_from_coords(lat = c(53.3703, 53.41145, 53.3096),
                                    lon = c(-1.47119, -1.500204, -1.478715),
                                    full_details = TRUE)
  expect_type(full_details, "list")
  expect_equal(length(full_details), 3)
  expect_identical(full_details[[1]]$words, "elite.icon.levels")
})

test_that("Errors are handled", {
  skip_if(identical(Sys.getenv("WTW_API_KEY"), ""),
          message = "API key environment variable not set.")
  invalid_coord_example <- suppressWarnings(words_from_coords(lat = 100, lon = -0.1266))
  expect_match(invalid_coord_example, "ERROR")
})


R/coords_from_words.R
#' Returns the coordinates for a given what3words address
#'
#' @param words A what3words location
#' @param full_details `FALSE` (default) to return only the coordinates.
#' `TRUE` for full details.
#'
#' @return If `full_details == FALSE` (the default) a matrix of WGS84
#' coordinates. Otherwise a list of full details for each what3words location.
#' @export
#'
#' @examplesIf !identical(Sys.getenv("WTW_API_KEY"), "")
#' coords_from_words("hotdog.jumping.frog")
coords_from_words <- function(words,
                              full_details = FALSE) {

  # Get API key from environment variable
  key <- whatthreewords::get_api_key()

  # Build requests - vectorised over coordinates
  requests <- mapply(function(words, key)
    httr2::request("https://api.what3words.com/v3/") |>
      httr2::req_url_path_append("convert-to-coordinates") |>
      httr2::req_url_query(words = words,
                           key = key)  |>
      httr2::req_user_agent("whatthreewords (https://github.com/DavidASmith/whatthreewords)") |>
      httr2::req_error(is_error = function(resp) FALSE),
    words = words,
    MoreArgs = list(key = key),
    SIMPLIFY = FALSE)


  # Make requests
  responses <-  requests |>
    lapply(httr2::req_perform)

  # Get JSON content of response
  contents <- responses |>
    lapply(httr2::resp_body_json)

  # Return coordinates or full details
  if(full_details) {
    out <- contents
  } else {
    out <- mapply(function(content, response)
    {if(response$status_code == 200) {
      matrix(c(content$coordinates$lat, content$coordinates$lng),
             nrow = 1)
    } else {
      msg <- paste0("ERROR: ", content$error$code, ": ", content$error$message)
      warning(msg, call. = FALSE)
      matrix(c(NA, NA), nrow = 1)
    }
    },
    content = contents,
    response = responses,
    SIMPLIFY = FALSE)

    out <- do.call(rbind, out)
    colnames(out) <- c("lat", "lon")
  }
  out
}


R/get_available_languages.R
#' Gets available languages for use in the what3words API
#'
#' @return A dataframe of available languages.
#' @export
#'
#' @examplesIf !identical(Sys.getenv("WTW_API_KEY"), "")
#' get_available_languages()
get_available_languages <- function() {

  # Get API key from environment variable
  key <- whatthreewords::get_api_key()

  # Build request
  request <- httr2::request("https://api.what3words.com/v3/") |>
    httr2::req_url_path_append("available-languages") |>
    httr2::req_url_query(key = key)  |>
    httr2::req_user_agent("whatthreewords (https://github.com/DavidASmith/whatthreewords)")

  # Make request
  response <- request |>
    httr2::req_perform()

  # Return dataframe
  content <- response |>
    httr2::resp_body_json()

  out <- content$languages |>
    lapply(data.frame) |>
    dplyr::bind_rows()
  out
}


R/words_from_coords.R
#' Get what3words address from coordinates
#'
#' @param lat Latitude
#' @param lon Longitude
#' @param language Code for the language of the returned words.
#' @param full_details Whether to return the full details from the API, or only
#' the what3words address for the coordinates.
#'
#' @return If `full_details` is FALSE (the default), returns a character vector
#' of the what3words for the submitted coordinates. Otherwise returns a list of
#' the full details returned by the API.
#' @export
#'
#' @examplesIf !identical(Sys.getenv("WTW_API_KEY"), "")
#' words_from_coords(lat = 51.5095, lon = -0.1266)
words_from_coords <- function(lat,
                              lon,
                              language = "en",
                              full_details = FALSE) {


  # Get API key from environment variable
  key <- whatthreewords::get_api_key()

  # Build requests - vectorised over coordinates
  requests <- mapply(function(lat, lon, language, key)
    httr2::request("https://api.what3words.com/v3/") |>
      httr2::req_url_path_append("convert-to-3wa") |>
      httr2::req_url_query(coordinates = paste0(lat,",",lon),
                           language = language,
                           key = key) |>
      httr2::req_user_agent("whatthreewords (https://github.com/DavidASmith/whatthreewords)") |>
      httr2::req_error(is_error = function(resp) FALSE),
    lat = lat,
    lon = lon,
    MoreArgs = list(language = language,
                    key = key),
    SIMPLIFY = FALSE)


  # Make requests
  responses <-  requests |>
    lapply(httr2::req_perform)

  # Get JSON content of response
  contents <- responses |>
    lapply(httr2::resp_body_json)

  # Return words or full details
  if(full_details) {
    if(length(contents) == 1) {
      out <- contents[[1]]
    } else {
      out <- contents
    }
  } else {
    out <-
      mapply(function(content, response)
      {if(response$status_code == 200) {
        content$words
      } else {
        msg <- paste0("ERROR: ", content$error$code, ": ", content$error$message)
        warning(msg, call. = FALSE)
        msg
      }
      },
      content = contents,
      response = responses,
      SIMPLIFY = FALSE) |>
      unlist()
  }
  out

}

